lib/elixir/lib/module/types/infer.ex

```diff
@@ -152,6 +131,121 @@ defmodule Module.Types.Infer do
     end
   end
 
+  defp unify_structs(left_pairs, right_pairs, stack, context) do
+    with {:ok, left_module} when is_atom(left_module) <- fetch_struct_pair(left_pairs),
+         {:ok, right_module} when is_atom(right_module) <- fetch_struct_pair(right_pairs) do
+      if left_module == right_module do
+        :ok
+      else
+        left = {:map, [{:required, {:atom, :__struct__}, left_module}]}
+        right = {:map, [{:required, {:atom, :__struct__}, right_module}]}
+        error({:unable_unify, left, right}, stack, context)
+      end
+    else
+      _ -> :ok
+    end
+  end
+
+  defp unify_maps(source_pairs, target_pairs, %{context: :pattern} = stack, context) do
+    source_pairs = expand_struct(source_pairs)
+    target_pairs = expand_struct(target_pairs)
+
+    # Since maps in patterns only support literal keys (excluding maps)
+    # we can do exact type match without subtype checking
```

Comment:
This is no longer true in OTP 23. I'm not sure Elixir supports it yet, though. Either way the comment should probably mention this

Reply:
Today we already support `^key` where `key` is a variable. So as long as this code handles `^expr` properly - which is by getting the type of `expr`, then it will just work when we do the changes in OTP 23.

In other words, this code should assume that `^expr` is a valid key and that `expr` can be anything. The compiler is the one restricting it to only variables, but this code doesn't have to. :)